src/sdft_interface.jl

```diff
@@ -0,0 +1,307 @@
+    nextiter = iterate(x, state.nextdatastate)
+    if nextiter === nothing
+        return nothing
+    end
+    nextdatapoint, nextdatastate = nextiter
+    dfthistory = state.dfthistory
+    fragment = state.fragment
+    n = windowlength(method)
+    # explicit fft of shifted fragment until dfthistory is complete
+    if dfthistory !== nothing && state.iteration <= maximum(dftback(method))
+        updatefragment!(fragment, nextdatapoint, state.iteration)
+        dft = shiftedfft(fragment, state.iteration)
+    else
+        dft = state.dft
+        statedata = StateData(dfthistory, fragment, nextdatapoint, n, state.iteration)
+        updatedft!(dft, x, method, statedata)
+        updatefragment!(fragment, nextdatapoint, state.iteration)
+    end
+    updatedfthistory!(dfthistory, dft, n, state.iteration + 1)
+    return SDFTState(dft, dfthistory, fragment, nextdatastate, state.iteration + 1)
+end
+
+function shiftedfft(x, delay)
+    n = length(x)
+    y = fft(x)
+    delayfactor = exp(2π*im*delay/n)
+    updatedfactor = one(delayfactor)
+    for k in eachindex(y)
+        y[k] *= updatedfactor
+        updatedfactor *= delayfactor
+    end
+    return y
+end
+
+function updatedfthistory!(dfthistory, dft, n, iteration)
+    offset = rem((iteration - 1) * n, length(dfthistory))
+    dfthistory[(1:n) .+ offset] .= dft
+end
+
+function updatedfthistory!(::Nothing, args...) end
+
+function updatefragment!(fragment, nextdatapoint, iteration)
+    n = length(fragment)
+    offset = rem(iteration - 1, n)
+    fragment[begin + offset] = nextdatapoint
+end
+
+function updatefragment!(::Nothing, ::Any, ::Any) end
+
+## Iterator
+
+struct SDFTIterator{M, T}
+    method::M
+    data::T
+    safe::Bool
+end
+
+getmethod(iterator::SDFTIterator) = iterator.method
+getdata(iterator::SDFTIterator) = iterator.data
+issafe(iterator::SDFTIterator) = iterator.safe
+
+Base.IteratorSize(::SDFTIterator{<:Any,T}) where T = IteratorSizeWrapper(Base.IteratorSize(T))
+IteratorSizeWrapper(::Base.HasShape) = Base.HasLength() # do not inherit shape
+IteratorSizeWrapper(iteratorsizetrait) = iteratorsizetrait # inherit everything else
+
+function Base.length(iterator::SDFTIterator)
+    method = getmethod(iterator)
+    data = getdata(iterator)
+    return length(data) - windowlength(method) + 1
+end
+
+Base.eltype(::SDFTIterator{M,T}) where {M,T} = Vector{Complex{eltype(T)}}
+
+Base.isdone(iterator::SDFTIterator) = Base.isdone(getdata(iterator))
+Base.isdone(iterator::SDFTIterator, state::SDFTState) = Base.isdone(getdata(iterator), state.nextdatastate)
+
+"""
+    sdft(method, x[, safe=true])
+
+Return an iterator to produce a sliding DFT of `x` using the given `method`.
+If `safe == true` (the default behavior), this iterator produces a new vector at each iteration.
+
+Set the optional argument `safe=false` to improve performance by reducing allocations,
+at the expense of unexpected behavior if the resulting vector is mutated between iterations.
+"""
```

Comment:
With `?sdft` the documentation seems very thin.
I think we need to add simple example here to make it clear.

But maybe we can also do the following:
```
(sdft::SDFT)(x, safe=true) = SlidingDFTs.SDFTIterator(sdft, x, safe)


julia> SDFT(10)(y)
FourierTools.SlidingDFTs.SDFTIterator{SDFT{Int64, ComplexF64}, Vector{Float32}}(SDFT{Int64, ComplexF64}(10, 0.8090169943749475 + 0.5877852522924731im), Float32[5.0, 4.045085, 1.545085, -1.545085, -4.045085, -5.0, -4.045085, -1.545085, 1.545085, 4.045085  …  -1.4697619f-14, 3.0, 4.4068714f-15, -3.0, 5.8838754f-15, 3.0, -1.6174623f-14, -3.0, 2.646537f-14, 3.0], true)

julia> SDFT(10)(y, false)
FourierTools.SlidingDFTs.SDFTIterator{SDFT{Int64, ComplexF64}, Vector{Float32}}(SDFT{Int64, ComplexF64}(10, 0.8090169943749475 + 0.5877852522924731im), Float32[5.0, 4.045085, 1.545085, -1.545085, -4.045085, -5.0, -4.045085, -1.545085, 1.545085, 4.045085  …  -1.4697619f-14, 3.0, 4.4068714f-15, -3.0, 5.8838754f-15, 3.0, -1.6174623f-14, -3.0, 2.646537f-14, 3.0], false)

julia> (sdft::SDFT)(x, safe=true) = F.SDFTIterator(sdft, x, safe)
```

In that way we only need one docstring.
Also, do we really need the submodule `SlidingDFTs`? So far we don't do any submodule and maybe it's easier to just include it normally.